feat: allow setting autoscaling options to deployed KService

zroubalik opened the pull request:
Signed-off-by: Zbynek Roubalik <zroubali@redhat.com>

Users can specify (autoscaling) options in `func.yaml`. The options are added/update/removed based on the content in `func.yaml` file.

If there's a demand, we can add these options as flags to `func deploy` command (same way they are exposed in `kn service create/update` -> with removal option), but I think that specification in config file is enough.


```yaml
name: test
namespace: ""
runtime: go
...
options:
  scale:
    min: 0
    max: 10
    metric: rps
    target: 75
    utilization: 75
```

`concurrency.limit` will be added in the following PR, which will handle resource limits/requests as proposed here: https://github.com/boson-project/func/pull/374#discussion_r649070413

Relates: #164
Fixes: #264

zroubalik commented on config.go:
```diff
@@ -28,6 +29,16 @@ type Env struct {
 }
 
+type Options struct {
+	ScaleMin              *int    `yaml:"scale-min,omitempty"`
+	ScaleMax              *int    `yaml:"scale-max,omitempty"`
+	ScaleInit             *int    `yaml:"scale-init,omitempty"`
+	AutoscaleWindow       *string `yaml:"autoscale-window,omitempty"`
+	ConcurrencyLimit      *int64  `yaml:"concurrency-limit,omitempty"`
+	ConcurrencyTarget     *int    `yaml:"concurrency-target,omitempty"`
+	ConcurrencyUtilizatin *int    `yaml:"concurrency-utilization,omitempty"`
```
I am not sure whether we should use the same notation as it is used in `kn` ie. dashes in the name or whether we should use CamelCase notation that's being used in the rest of the file...

lkingland replied:
tl;dr  I see no compelling reason to maintain exact equivalence, and lean towards CamelCase at a minimum.

My first thought on reading this was much the same:  Glad we have it; a bit less enthusiastic about the naming.  I would even go a little further... Following is perhaps overly opinionated, but why not share!

I am sure that many would encourage us to maintain homogeneity between the two.  However, I think this actually leads to a couple challenges:  First, it encourages a false equivalence between the systems, causing confusion by users and developers alike; especially when we do deviate.  Second, we risk grandfathering in naming which, with the benefit of hindsight, could be easily improved or made to fit our use case a bit better.  I see the argument for exact equivalence as valid, but weaker than either:  "it's easier to remember if you're familiar with both systems".

We have a golden and ephemeral opportunity to lay down consistent, coherent naming now, before we are restrained by the iron yoke backwards compatibility.  I say we take it whenever the change is a clear improvement and not just bike-shedding.

For instance, would anyone argue that this is clearer semantics?
```
scaling:
  initial: 2
  min: 1
  max:  10
  window: 10s
concurrency:
  target: 75
  utilization: 75
  limit: 100
```
(This is of course just a conversational example more than an actual suggestion; illustrating a point.  I like the practical original suggestion of adopting CamelCase just fine)

lance replied:
I actually like your "conversational example" a bit more than using an `options` keyword. My first thought upon seeing the PR was, "why an `options` key" instead of just having `scale-min`, for example, at the top level? I suppose there is benefit in the fact that, as new flags/options are added they can all just be stuffed into one `options` bag instead of being scattered across the file. I do agree, however, that your example, @lkingland is more readable/understandable. What about something like this (also a conversational example, but maybe also a suggestion :shrug:)

```yaml
options:
  scaling:
    initial: 2
    min: 1
    max:  10
    window: 10s
  concurrency:
    target: 75
    utilization: 75
    limit: 100
```

zroubalik replied:
Yeah, I created the `options` exactly for this, to not have all the settings scattered across the func.yaml file.

I like this suggestion, will refactor the config this way, thanks a lot for the input.

lance commented on config.go:
```diff
@@ -166,7 +189,7 @@ func writeConfig(f Function) (err error) {
 
 // validateVolumes checks that input Volumes are correct and contain all necessary fields.
-// Returns array of error messages, empty if none
+// Returns array of error messages, empty if none errors are found
```
```suggestion
// Returns array of error messages, empty if no errors are found
```
Such a tiny nit here and below...

zroubalik replied:
Thx!

lance commented on config.go:
```diff
@@ -28,6 +29,16 @@ type Env struct {
+	AutoscaleWindow       *string `yaml:"autoscale-window,omitempty"`
+	ConcurrencyLimit      *int64  `yaml:"concurrency-limit,omitempty"`
+	ConcurrencyTarget     *int    `yaml:"concurrency-target,omitempty"`
+	ConcurrencyUtilizatin *int    `yaml:"concurrency-utilization,omitempty"`
```
```suggestion
	ConcurrencyUtilization *int    `yaml:"concurrency-utilization,omitempty"`
```
Spelling nit

lance commented on config.go:
```diff
@@ -241,7 +264,27 @@ func ValidateEnvs(envs Envs) (errors []string) {
+
+// validateOptions checks that input Options are correctly set.
+// Returns array of error messages, empty if none errors are found
+func validateOptions(options Options) (errors []string) {
```
Why no validation for the other options? Seems like you could check them, for example, similar to this.

- `minScale >= 0`
- `maxScale >= minScale`
- `initialScale >= minScale`
- `0s <= window <= 1h`
- `utilization <= 100`

lance commented on docs/guides/func_yaml.md:
````diff
@@ -62,6 +62,27 @@ volumes:
 ```
 
+### `options`
+Options allows you to set specific configuration for the deployed funciton, ie. it allows you to tweak Knative Service options related to autoscaling and other properties, if these options are not set in the default ones will be used. This options are based on the ones provided by `kn service` [command](https://github.com/knative/client/blob/main/docs/cmd/kn_service_create.md#options).
````
```suggestion
Options allows you to set specific configuration for the deployed funciton, allowing you to tweak Knative Service options related to autoscaling and other properties. If these options are not set, the Knative defaults will be used. 
```

Minor wordsmithing, and I think it might be best to avoid referencing the `kn` options, just to avoid confusion among users - but I'm not firm on this suggestion.

lance replied:
Oh also a typo... `funciton`.

lance commented on docs/guides/func_yaml.md:
```diff
@@ -62,6 +62,27 @@ volumes:
+- `autoscale-window`: Duration to look back for making auto-scaling decisions. The Service is scaled to zero if no request was received in during that time. (eg: 10s)
+- `concurrency-limit`: Hard Limit of concurrent requests to be processed by a single replica.
+- `concurrency-target`: Recommendation for when to scale up based on the concurrent number of incoming request. Defaults to `concurrency-limit` when given.
+- `concurrency-utilization`: Percentage of concurrent requests utilization before scaling up.
```
It might be beneficial to have the default values listed as well.

zroubalik replied:
I was thinking about documenting default values, but those are based on a entries in a config in knative-serving namespace, so they could vary across individual installations.

lance reviewed (commented):
In general looks great! A few minor comments/suggestions.

zroubalik commented:
@boson-project/core  PTAL, I have refactored this PR per feedback, thanks for that!

@markusthoemmes suggested that we should dropp `autoscale window` and `init scale` as those are not yet stable and might break, I have introduced `scale metric` instead.

markusthoemmes commented on config.go:
```diff
@@ -28,6 +28,23 @@ type Env struct {
+	Metric *string `yaml:"metric,omitempty"`
+}
+
+type ConcurrencyOptions struct {
```
With the introduction of `metric` above, is this really still limited to `concurrency`? I'd guess only `Limit` is?

zroubalik replied:
How would you map those values into the config then so it is not bloated and easy to read? I am open to any suggestions :)

markusthoemmes replied:
maybe `target` and `utilization` should move into the scale options (that's what they are after all), and the concurrency limit should get a special place somehow (as it also is in Knative: It's inferred as an autoscaling setting (inferred into `target` actually) but also significant as a routing and traffic behavior setting)

rhuss replied:
Actually, I think one design goal would be to allow to map the config names here directly to `kn` options. Unfortunately, we have chosen `--concurrency-target` and `--concurrency-utilization`, so its not so easy to move those around without confusing this UX. Of course we could do this hand-in-hand with a corresponding change on `kn`, but this would be a more tricky because of backwards compatibility issues ...

lance replied:
> Actually, I think one design goal would be to allow to map the config names here directly to `kn` options. Unfortunately, we have chosen `--concurrency-target` and `--concurrency-utilization`, so its not so easy to move those around without confusing this UX.

```yaml
options:
  scaling:
    initial: 2
    min: 1
    max:  10
    window: 10s
  concurrency:
    target: 75
    utilization: 75
    limit: 100
```

Have we not already made enough of a leap with the YAML in this format to assume the user experience does not have a direct correlation to `kn` without at least a little interpretation? The function development experience is meant to be clear, concise, opinionated, and targeted towards developers new to Knative. Does a strict adherence to `kn` command semantics necessarily help this goal? I'm not sure. :shrug: 

I know that, as a Linux user I can never remember if I should use `adduser` or `useradd`, so I understand the UX concern! :smile: 

> Of course we could do this hand-in-hand with a corresponding change on kn, but this would be a more tricky because of backwards compatibility issues ...

What if we think of `func` (and any other `kn` plugin, I suppose) as testing ground for user experiences that __may__ make their way into `kn`, with proper deprecation timelines and `--experimental` flags? So that instead of needing to make a hand-in-hand corresponding change, there is a process for `kn` in adopting features/usability/etc from plugins.

markusthoemmes replied:
Ack, I don't think we need to inherit all design choices from `kn` (or even Knative!) if we can do better.

zroubalik replied:
@markusthoemmes so you'd like to have the config like this?
```
name: test
namespace: ""
runtime: go
...
options:
  scale:
    min: 0
    max: 10
    metric: rps
    target: 75
    utilization: 75
  concurrency:
    limit: 100
```

lkingland replied:
> Ack, I don't think we need to inherit all design choices from `kn` (or even Knative!) if we can do better.

We can do better indeed!  We have the benefits of both hindsight and a higher level of abstraction, thanks in no small part to kn and Knative in general.  I most certainly like the clarity and simplicity of these options as they are shaping up here.

rhuss replied:
Ok, fine with it and let's fix in on `kn` and also support other types of scaling then only concurrency (rps, cpu)

Just to be sure about this particular change: `concurrency.limit` is really only about the concurrency while `scale.target` is about any of concurrency, rps or cpu depending of which scaling type has been selected with the annotation `autoscaling.knative.dev/metric`. How would one specify "hard limits" then for RPS or CPU based scaling ? (if we stay with the analogy of soft and hard limits for target/limit). Is it also via `...spec.containerConcurrency` (then that field would be misnamed) or is this just not be possible ? If it is the former case, then I would put the `concurrency.limit` also to the scale group, as it feels unnatural to have target and limit in different sections.

rhuss replied:
> Ok, fine with it and let's fix in on kn and also support other types of scaling then only concurrency (rps, cpu)

Introducing the scaling metric (`--scale-metric`) is then also a good hook for changing the `kn` option name to `--scale-target`.

I still believe having user-faced wording aligned across abstractions is a good thing.

markusthoemmes replied:
There are no hard limits for RPS. CPU limits would be set via the container's resource block.

I would not conflate those though. The concurrency limit is not just impacting scale, but it's also impacting traffic behavior. Likewise, CPU-limits are not only impacting scale but also the behavior of the container. Target on the other hand, is strictly a scaling concern.

zroubalik replied:
I have moved `target` and `utilization` to `scale` section in https://github.com/boson-project/func/pull/374/commits/7a8b7ef6bcf4a13c1f6858190b7db8fd8dedafa6

rhuss replied:
Ok, I understand. Then I wonder why not then grouping all limits together ? Like something as:

```yaml
resources:
  limits:
    cpu: ....
    memory: ....
    concurrency: ....
  requests:
    cpu: ...
    memory: ...
```

@zroubalik How do you specify the resource limits currently for the container ?

zroubalik replied:
Resource requests/limits are not being exposed at the moment. But my idea was to add them once we agree on the scaling options and my intention was to add them to the options exactly the way as you proposed. The concurrency added to limits seems nice! Great idea.

markusthoemmes replied:
Nice idea indeed! I like!

zroubalik replied:
I have removed `concurrency.limits` from this PR and will handle that in another one together with resource requests/limits.

lkingland reviewed (approved):
Well, I like it.
